File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ColorSelector.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonList;
import com.urbanairship.json.JsonMap;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class ColorSelector {
    @Nullable
    private final Platform platform;
    private final boolean darkMode;
    @ColorInt
    private final int color;

    public ColorSelector(@Nullable Platform platform, boolean darkMode, int color) {
        this.platform = platform;
        this.darkMode = darkMode;
        this.color = color;
    }

    @NonNull
    public static ColorSelector fromJson(@NonNull JsonMap json) throws JsonException {
        String platformString = json.opt("platform").optString();
        Platform platform = platformString.isEmpty() ? null : Platform.from(platformString);
        boolean darkMode = json.opt("dark_mode").getBoolean(false);
        JsonMap colorJson = json.opt("color").optMap();
        @ColorInt Integer color = HexColor.fromJson(colorJson);
        if (color == null) {
            throw new JsonException("Failed to parse color selector. 'color' may not be null! json = '" + json + "'");
        }

        return new ColorSelector(platform, darkMode, color);
    }

    @NonNull
    public static List<ColorSelector> fromJsonList(@NonNull JsonList json) throws JsonException {
        List<ColorSelector> selectors = new ArrayList<>(json.size());
        for (int i = 0; i < json.size(); i++) {
            JsonMap selectorJson = json.get(i).optMap();
            ColorSelector selector = ColorSelector.fromJson(selectorJson);
            selectors.add(selector);
        }
        return selectors;
    }

    @Nullable
    public Platform getPlatform() {
        return platform;
    }

    public boolean isDarkMode() {
        return darkMode;
    }

    @ColorInt
    public int getColor() {
        return color;
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ColorSelector.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import androidx.annotation.ColorInt
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonList
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonValue
import com.urbanairship.json.optionalField
import com.urbanairship.json.optionalMap

public class ColorSelector public constructor(
    public val platform: Platform?,
    public val isDarkMode: Boolean,
    @field:ColorInt public val color: Int
) {

    public companion object {
        private const val KEY_PLATFORM = "platform"
        private const val KEY_DARK_MODE = "dark_mode"
        private const val KEY_COLOR = "color"

        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): ColorSelector {
            val content = json.requireMap()

            @ColorInt val color = HexColor.fromJson(content.optionalMap(KEY_COLOR))
            if (color == null) {
                throw JsonException("Failed to parse color selector. 'color' may not be null! json = '$json'")
            }

            return ColorSelector(
                platform = content[KEY_PLATFORM]?.let(Platform::from),
                isDarkMode = content.optionalField(KEY_DARK_MODE) ?: false,
                color = color
            )
        }

        @Throws(JsonException::class)
        public fun fromJsonList(json: JsonList): List<ColorSelector> {
            return json
                .map(::fromJson)
                // Ignore any non-android selectors.
                .filter { it.platform == null || it.platform == Platform.ANDROID }
        }
    }
}
